src/scheduler/Scheduler.cpp

```diff
@@ -226,203 +226,179 @@ std::vector<std::string> Scheduler::callFunctions(
+            int available = slots - thisHostResources.usedslots();
+            available = std::max<int>(available, 0);
 
-                // Provided we're not executing threads, execute the functions
-                // now
-                if (!isThreads) {
-                    funcQueue->enqueue(msg);
-                    executed.at(nextMsgIdx) = thisHost;
-                    addFaaslets(msg);
-                }
+            // Claim as many as we can
+            nLocally = std::min<int>(available, nMessages);
+        }
+
+        // Handle those that can be executed locally
+        if (nLocally > 0) {
+            logger->debug(
+              "Executing {}/{} {} locally", nLocally, nMessages, funcStr);
+            for (int i = 0; i < nLocally; i++) {
+                localMessageIdxs.emplace_back(i);
+                executed.at(i) = thisHost;
             }
+        }
 
-            // If some are left, we need to distribute
-            if (remainder > 0) {
-                // At this point we have a remainder, so we need to distribute
-                // the rest Get the list of registered hosts for this function
-                std::unordered_set<std::string>& thisRegisteredHosts =
-                  registeredHosts[funcStr];
-
-                // Schedule the remainder on these other hosts
-                for (auto& h : thisRegisteredHosts) {
-                    int nOnThisHost =
-                      scheduleFunctionsOnHost(h, req, executed, nextMsgIdx);
-
-                    remainder -= nOnThisHost;
-                    if (remainder <= 0) {
-                        break;
-                    }
+        // If some are left, we need to distribute
+        int offset = nLocally;
+        if (offset < nMessages) {
+            // At this point we have a remainder, so we need to distribute
+            // the rest over the registered hosts for this function
+            std::unordered_set<std::string>& thisRegisteredHosts =
+              registeredHosts[funcStr];
+
+            // Schedule the remainder on these other hosts
+            for (auto& h : thisRegisteredHosts) {
+                int nOnThisHost =
+                  scheduleFunctionsOnHost(h, req, executed, offset);
+
+                offset += nOnThisHost;
+                if (offset >= nMessages) {
+                    break;
                 }
             }
+        }
 
-            // Now we need to find hosts that aren't already registered for
-            // this function and add them
-            if (remainder > 0) {
-                std::unordered_set<std::string>& thisRegisteredHosts =
-                  registeredHosts[funcStr];
-
-                std::unordered_set<std::string> allHosts = getAvailableHosts();
+        if (offset < nMessages) {
+            // At this point we know we need to enlist unregistered hosts
+            std::unordered_set<std::string> allHosts = getAvailableHosts();
```

Comment:
If we were to use an ordered container like `std::set` instead of `std::unordered_set` we could use [`std::set_difference`](https://en.cppreference.com/w/cpp/algorithm/set_difference), which feels a bit like what we are doing.

Reply:
Yes that might make sense here. In general I always avoid the ordered collections as the performance of `std::map` is noticeably different to `std::unordered_map` when doing inserts. In this case the set will be static so that would be good.